Add option `--output` to `crystal docs` generator

straight-shoota opened the pull request:
Allows to configure output directory for `crystal docs` as command line argument.
Fixes #4918 in part.

There is also the debate whether the default location should be `doc` (current) or `docs`. A change could be incorporated into this PR if desired, but discussion should take place in #4918.

asterite commented:
More options usually means the tool couldn't find a standard way to do it. What's wrong with doc?

straight-shoota commented:
@asterite Could we keep the discussion about the general idea in #4918 please? There are already some arguments for making it configurable. `doc` just won't fit for everyone and there is little sense in forcing everyone when it is no big deal to customize.

asterite commented on src/compiler/crystal/tools/doc/generator.cr:
```diff
@@ -28,7 +28,7 @@ class Crystal::Doc::Generator
   }
 
-  def initialize(@program : Program, @included_dirs : Array(String), @dir = "./doc")
+  def initialize(@program : Program, @included_dirs : Array(String), @dir = "./docs")
```
Maybe don't provide a default here, since it's also in `command/docs.cr`

straight-shoota replied:
Yeah this duplication is bad. But you'd also want the default value if the generator is directly used from code.
I was thinking about adding a dedicated struct for holding option values. This could be initialized (with implicit defaults) instead of `output_directory = "./docs"` and be used as the second argument for `Doc::Generator.new`.
```crystal
struct Crystal::Doc::Generator::Options
  property included_dirs = [] of String
  property output_dir = "./docs"
end

asterite replied:
Or set the default as `nil` and then do `@dir ||= "./docs"`. And in the option parser place initialize it to `nil`.

straight-shoota replied:
This feels dirty :P I'm working on an options container... ;)

asterite replied:
Oh, I meant:

```crystal
def initialize(dir = nil)
  @dir = dir || "./docs"
end
```

I don't think that's dirty at all.

Or remove the default value. The doc generator isn't used outside the compiler, and if it is, passing a mandatory directory string should be fine.

Adding an extra type complicates the design.

straight-shoota replied:
I really like including the default value in the help message: `Set the output directory (default: ./docs)` For this to work, we'll need the value outside of the Generator constructor. So no default value there.

I feel like a dedicated options type is a cleaner solution on the long run. But I'm not fixed on that.

straight-shoota commented:
@asterite I removed the default value for `@dir` in the previous commit. And I also added a commit that introduces an options object, to show how it could be done (but doesn't need to).

bew commented on src/compiler/crystal/tools/doc/generator.cr:
```diff
@@ -1,7 +1,13 @@
 class Crystal::Doc::Generator
+  class Options
+    property included_dirs = [] of String
+    property output_dir = "./docs"
```
I think you can remove the `./`

straight-shoota replied:
It was in there before, so I just copied it. But you're probably right, I don't see any reason as well.

Sija commented on src/compiler/crystal/tools/doc/generator.cr:
```diff
@@ -72,15 +77,15 @@ class Crystal::Doc::Generator
     end
 
-    File.write "#{@dir}/index.html", MainTemplate.new(body, types, repository_name)
+    File.write "#{@options.output_dir}/index.html", MainTemplate.new(body, types, repository_name)
```
It'd be better to use (thinking about windows compat):

```cr
File.join(@options.output_dir, "index.html")
```

Sija commented on src/compiler/crystal/tools/doc/generator.cr:
```diff
@@ -72,15 +77,15 @@ class Crystal::Doc::Generator
 
   def copy_files
-    Dir.mkdir_p "#{@dir}/css"
-    Dir.mkdir_p "#{@dir}/js"
```
ditto

Sija commented on src/compiler/crystal/tools/doc/generator.cr:
```diff
@@ -72,15 +77,15 @@ class Crystal::Doc::Generator
 
-    File.write "#{@dir}/css/style.css", StyleTemplate.new
-    File.write "#{@dir}/js/doc.js", JsTypeTemplate.new
+    File.write "#{@options.output_dir}/css/style.css", StyleTemplate.new
```
ditto

straight-shoota commented:
What are others' opinions on storing options in a dedicated type? (see [line comment thread](https://github.com/crystal-lang/crystal/pull/4937#pullrequestreview-61566243))
> I feel like a dedicated options type is a cleaner solution on the long run. But I'm not fixed on that.
I'm happy to address the issues regarding file paths once we're settled on this.

straight-shoota commented on src/compiler/crystal/tools/doc/generator.cr:
```diff
@@ -1,7 +1,13 @@
 class Crystal::Doc::Generator
+  class Options
```
If we keep this, it could be a struct (using record macro) to reduce memory impact. Naming could also be improved, maybe `GeneratorConfig`?

straight-shoota commented:
@asterite WDYT?

asterite commented:
@straight-shoota Cool, thanks! 🙇
